src/script/interpreter.cpp

```diff
@@ -469,8 +469,13 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
                 }
             }
 
-            if (opcode == OP_CAT ||
-                opcode == OP_SUBSTR ||
+            if (opcode == OP_CAT) {
+                if (sigversion == SigVersion::BASE || sigversion == SigVersion::WITNESS_V0) {
```

Comment:
why not move this check down to the actual `OP_CAT` switch clause?

Reply:
I think that's a hard fork? Currently `0 IF CAT ENDIF 1` results in failure (CAT is disabled even in unexecuted branches), but moving the check into the switch would mean the check is only made in executed branches (since the switch is conditional on the value of `fExec`).

Reply:
How can that be? This clause is only executed during script execution, and moving it later in the loop shouldn't really make a difference as far as I can tell.

The script success behavior you are describing is done in the pre-processing step I believe.

Reply:
The `for(..., ++opcode_pos)` iterates over every opcode, the `bool fExec` step determines if each opcode will be executed, but the only thing conditional on `fExec` is the `stack,push_back()` and `switch` statements below. The disabled opcode test in particular is not conditional on `fExec`.

Reply:
Ah, yeah I think you're right. 

Since behavior should stay the same for non-tapscript execution (and OP_SUCCESS is not a thing in that case), the pre-processsing success check is irrelevant.